### PR TITLE
UCM/UCS/RCACHE: allow rcache to register for multiple events

### PR DESCRIPTION
**Pull request by bureddy**
- detached VM_UNMAP from cudaFree
        - allow rcache to register with multiple events

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7404/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7405/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7406/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7407/ for details (Mellanox internal link).

File: src/ucs/sys/rcache.h [hunk 1]
```diff
@@ -106,6 +106,7 @@ struct ucs_rcache_params {
                                                      Must be smaller or equal to
                                                      system page size. */
     size_t                 max_alignment;       /**< Maximum alignment */
+    int                    ucm_events;          /**< UCM events to register */
```
Comment (yosefe) [inline src/ucs/sys/rcache.h]: currently  UCM_EVENT_VM_UNMAPPED and  UCM_EVENT_MEM_TYPE_FREE are supported

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -1164,6 +1166,10 @@ uct_ib_md_parse_reg_methods(uct_ib_md_t *md, const uct_ib_md_config_t *md_config
             rcache_params.region_struct_size = sizeof(uct_ib_rcache_region_t);
             rcache_params.alignment          = md_config->rcache.alignment;
             rcache_params.max_alignment      = ucs_get_page_size();
+            rcache_params.ucm_events         = UCM_EVENT_VM_UNMAPPED;
+            if (md_attr->cap.reg_mem_types & UCS_BIT(UCT_MD_MEM_TYPE_CUDA)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: maybe we want to test if there is any flag **except** host?

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -1431,17 +1438,17 @@ uct_ib_md_open(const char *md_name, const uct_md_config_t *uct_md_config, uct_md
...
     }
 
-    status = uct_ib_md_parse_reg_methods(md, md_config);
+    status = uct_md_query(&md->super, &md_attr);
     if (status != UCS_OK) {
         goto err_destroy_umr_qp;
     }
 
-    status = uct_ib_md_parse_device_config(md, md_config);
+    status = uct_ib_md_parse_reg_methods(md, &md_attr, md_config);
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: do we need to also change the order of destruction (in both error flow and good flow)

Comment (bureddy) [reply src/uct/ib/base/ib_md.c]: din't change the order here.  just moved uct_md_quey() up before uct_ib_md_parse_reg_methods().

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7411/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@yosefe addresses review comments. can you please review again?

**Pull request merged**